### PR TITLE
Add facilities for signal throttling and debouncing

### PR DESCRIPTION
**Pull request by phyBrackets**
Implementing #44

### CODE REVIEW

File: tests/signal/tst_signal.cpp [hunk 1]
```diff
@@ -93,6 +93,36 @@ TEST_CASE("Signal connections")
...
+        Signal<int> signal;
+        int count = 0;
+        auto handle = signal.connectWithDebouncing([&count](int value) {
+            count++;
+        }, 100);
+
+         signal.emit(1);
+         REQUIRE(count == 0);  // Debouncing interval hasn't passed
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Wait for debouncing interval
```
Comment (LeonMatthesKDAB) [inline tests/signal/tst_signal.cpp:121]: We definitely do not want to be waiting and rely on time this much in a test...
There are basically two things that are problematic:
1. we have to wait 100ms every time we run this test, if we want to add more debouncing tests, this means more waiting for tests to finish.
2. If the program is interrupted by the OS between `connectWithDebouncing` and `signal.emit` for more than 100ms (or whatever time we choose here), the test will fail as the timer will have expired before we thought. With such a long delay, this is highly unlikely, but for reasons against such a long delay, see `1.`, the shorter we choose this delay, the more likely this issue will be however.

The only real solution I can think of is mocking/faking time for our testing.
Generally this means to include some facility to override time, e.g. by having an optional<function<>> that we can set during testing to return any time we want, or by providing means to finish timers early.

A more high-level approach may be something like `libfaketime` (see Zeno's Edu-Report "Speeding up time based unit tests with libfaketime").
Though this is likely overkill for our use-case.

**Review by LeonMatthesKDAB — CHANGES_REQUESTED** [review]
Overall, a good step in the right direction.

I do believe that using Debouncer and Throttler objects would go a long way to simplify the API though.
In the end what basically happens is that a closure is constructed inside the function, which behaves just like an object, it just uses captured variables as members. (fun fact: this is actually how to construct an "object" in LISP). So this could just as well happen outside the function, so the Signal itself doesn't have to be concerned with this.

There are also a few implementation issues that I've noted.

File: tests/signal/tst_signal.cpp (hunk 1 above)
Comment (phyBrackets) [reply tests/signal/tst_signal.cpp]: Ah yes, that's true faking the time for testing could be one option. I'll look in that

File: src/kdbindings/signal.h [hunk 2]
```diff
@@ -242,6 +244,49 @@ class Signal
...
+                    std::this_thread::sleep_for(throttleDelay);
+
+                    shouldWait = false;
+                }).join();
+            };
+            return m_connections.insert({ throttleCallBack });
+        }
+
+        // Connect a callback function with Debouncing
+        Private::GenerationalIndex connectWithDebouncing(std::function<void(Args...)> const &slot, int interval)
```
[the remark this answers is not in the document]
Comment (phyBrackets) [reply src/kdbindings/signal.h:265]: Ah indeed it is challenging to implement debouncer and throttler in our current implementation , so basically if i am understanding clearly, we need a way that whenever a signal is received, we schedule a task to run after the debounce interval. If another signal is received before the interval expires, we cancel/remove the previously scheduled task and schedule a new one. And something similar with throttler ? We will need a mechanism to queue up the task/slots for debouncer and throttler, or we may use ConnectionEvaluator but need to look once the patch for it lands.

**Comment by LeonMatthesKDAB** [comment]
As this is rather stale now, I'll close the PR for now.
In case this becomes active again, feel free to re-open.

**Pull request closed without merging**